### PR TITLE
Constant propagation of delete() for StringBuilder and StringBuffer

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 621eb464d153c465a3d1c203bfd2cad7b6a31fae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129262353
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: jbmc/regression/jbmc-strings/ConstantEvaluationStringBufferDelete/Test.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  {
+    StringBuffer sb = new StringBuffer("abc");
+    sb.delete(0, index);
+    assert sb.toString().equals("ac");
+  }
+
+  public void testIndexOutOfBounds1()
+  {
+    StringBuffer sb = new StringBuffer("abc");
+    sb.delete(-1, 3);
```
Comment (smowton) [inline jbmc/regression/jbmc-strings/ConstantEvaluationStringBufferDelete/Test.java:55]: Normally I'd expect this to throw and therefore the assertion is unreachable; are you checking for a thrown exception here or do we deviate from the real StringBuilder's behaviour?

**Review by smowton — APPROVED** [review]

File: jbmc/regression/jbmc-strings/ConstantEvaluationStringBufferDelete/Test.java (hunk 1 above)
Comment (danpoe) [reply jbmc/regression/jbmc-strings/ConstantEvaluationStringBufferDelete/Test.java:55]: Yes this throws an exception. I'll remove the assertion and check specifically that an exception is thrown.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 76b89d56e9aa093eeda4b4d8401741cc6b37e9c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129299371
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**